File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top, bottom = 0, len(matrix)-1
        n = len(matrix[0])-1

        while top <= bottom:
            mid = (top + bottom)//2

            if(target >= matrix[mid][0] and target <= matrix[mid][-1] ):
                li ,ri = 0, len(matrix[mid])-1

                while li <= ri:
                    midi = (li + ri)//2
                    if(target == matrix[mid][midi]):
                        return True
                    elif(target > matrix[mid][midi]):
                        li = midi + 1
                    elif(target < matrix[mid][midi]): 
                        ri = midi - 1
                break

            elif(target > matrix[mid][-1]):
                top = mid + 1
            elif(target < matrix[mid][0]):
                bottom = mid - 1

        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
    let m = matrix.length
    let n = matrix[0].length

    let ol = 0
    let or = m - 1
    while (ol <= or) {
        let omid = Math.floor((ol + or) / 2)
        if (target >= matrix[omid][0] && target <= matrix[omid][n - 1]) {
            // do bin search here
            let l = 0
            let r = n - 1
            while (l <= r) {
                let mid = Math.floor((l + r) / 2)
                if (matrix[omid][mid] === target) return true

                if (target < matrix[omid][mid]) r = mid - 1;
                else l = mid + 1;
            }
            return false
        } else {
            if (target < matrix[omid][0]) or = omid - 1
            else if (target > matrix[omid][n - 1]) ol = omid + 1
        }
    }
    return false

};